File: LeetCode/CPP/287.find-the-duplicate-number.cpp
/*
 * @lc app=leetcode id=287 lang=cpp
 *
 * [287] Find the Duplicate Number
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution
{
public:
    int findDuplicate(vector<int> &nums)
    {
        int xor_val = 1;
        int ans = 0;
        for (int num : nums)
        {
            ans = ans xor num;
        }
        for (int i = 1; i < nums.size(); i++) {
            ans = ans xor i;
        }
        return ans;
    }
};
// @lc code=end

File: LeetCode/Python/287.find-the-duplicate-number.py
#
# @lc app=leetcode id=287 lang=python3
#
# [287] Find the Duplicate Number
#
from typing import List

# @lc code=start

# Brute -> Sort the array, then check for adjacent element (Modifies Array)
# Brute 2 -> Use a dictionary (Extra Space)
# Optimal: Floyd's Hare and Tortoise Algorithm, TODO: Read it.
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast = slow = slow2 = 0

        while True:
            fast = nums[nums[fast]]
            slow = nums[slow]

            if fast == slow:
                break

        while slow != slow2:
            slow = nums[slow]
            slow2 = nums[slow2]

        return slow

# @lc code=end

